src/query/pools/query_pools.go

```diff
@@ -50,47 +47,29 @@ const (
 
 // BuildWorkerPools builds a worker pool
 func BuildWorkerPools(
-	cfg config.Configuration,
-	logger *zap.Logger,
+	instrumentOptions instrument.Options,
 	scope tally.Scope,
-) (pool.ObjectPool, xsync.PooledWorkerPool, instrument.Options, error) {
-	workerPoolCount := cfg.DecompressWorkerPoolCount
-	if workerPoolCount == 0 {
-		workerPoolCount = defaultWorkerPoolCount
-	}
-
-	workerPoolSize := cfg.DecompressWorkerPoolSize
-	if workerPoolSize == 0 {
-		workerPoolSize = defaultWorkerPoolSize
-	}
-
-	instrumentOptions := instrument.NewOptions().
-		SetZapLogger(logger).
-		SetMetricsScope(scope.SubScope("series-decompression-pool"))
-
-	poolOptions := pool.NewObjectPoolOptions().
-		SetSize(workerPoolCount).
-		SetInstrumentOptions(instrumentOptions)
-
-	objectPool := pool.NewObjectPool(poolOptions)
-	objectPool.Init(func() interface{} {
-		workerPool := xsync.NewWorkerPool(workerPoolSize)
-		workerPool.Init()
-		return workerPool
-	})
-
-	writePoolSize := cfg.WriteWorkerPoolSize
-	if writePoolSize == 0 {
-		writePoolSize = defaultWorkerPoolCount
+	readPoolPolicy, writePoolPolicy xconfig.WorkerPoolPolicy,
+) (xsync.PooledWorkerPool, xsync.PooledWorkerPool, error) {
+	opts, readPoolSize := readPoolPolicy.Options()
+	instrumentOptions.SetMetricsScope(scope.SubScope("read-worker-pool"))
```

Comment:
instrumentOptions is immutable so you need to do:

```golang
readPoolIOpts = instrumentOptions.SetMetricsScope(scope.SubScope("read-worker-pool"))
opts = opts.SetInstrumentOptions(readPoolIOpts)
```

Reply:
Nice catch, just inlined it instead :)